src/bson.c

```diff
@@ -767,14 +796,66 @@ static const bson_visitor_t php_bson_visitors = {
 	{ NULL }
 };
 
+static bool map_element_matches_field_path(php_phongo_field_path_map_element* map_element, php_phongo_field_path* current)
+{
+	size_t i;
+
+	if (map_element->entry->size != current->size) {
+		return false;
+	}
+	for (i = 0; i < current->size; i++) {
+		if (strcmp(map_element->entry->elements[i], "$") == 0) {
+			continue;
+		}
+		if (strcmp(map_element->entry->elements[i], current->elements[i]) != 0) {
+			return false;
+		}
+	}
+	return true;
+}
+
+static php_phongo_field_path_map_element* map_find_field_path_entry(php_phongo_bson_state* state)
+{
+	size_t i;
+
+	/* Loop over all field path mappings, and for each, try to see whether it matches the current path */
+	for (i = 0; i < state->map.field_paths.size; i++) {
+		if (map_element_matches_field_path(state->map.field_paths.map[i], state->field_path)) {
+			return state->map.field_paths.map[i];
+		}
+	}
+	return NULL;
+}
+
+static void php_phongo_handle_field_path_entry_for_compound_type(php_phongo_bson_state* state, php_phongo_bson_typemap_types* type, zend_class_entry** ce)
+{
+	php_phongo_field_path_map_element* entry = map_find_field_path_entry(state);
+
+	if (entry) {
+		switch (entry->node_type) {
+			case PHONGO_TYPEMAP_NATIVE_ARRAY:
+			case PHONGO_TYPEMAP_NATIVE_OBJECT:
+				*type = entry->node_type;
+				break;
+			case PHONGO_TYPEMAP_CLASS:
+				*type = entry->node_type;
+				*ce   = entry->node_ce;
+				break;
+			default:
+				/* Do nothing - pacify compiler */
+				break;
```

Comment:
Is `php_phongo_bson_state state` always initialized with `PHONGO_BSON_STATE_INITIALIZER` (and layered with the default type map) before `php_phongo_handle_field_path_entry_for_compound_type()` is called at each level? I just want to make sure there is no concern with this "do nothing" behavior inadvertently leaving a previously applied field path applied to the state.

Would it make sense to revert to the default type map before calling this function?

Reply:
Yes, it is. We create a new state for each nested level, which is assigned with ``php_phongo_bson_state state = PHONGO_BSON_STATE_INITIALIZER;`` in ``php_phongo_bson_visit_document`` and ``php_phongo_bson_visit_array``: https://github.com/mongodb/mongo-php-driver/pull/793/files#diff-f7d29a680148f52d6601f59ed787f577R861